oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/function/client/OAuth2ClientHttpRequestInterceptor.java

```diff
@@ -0,0 +1,462 @@
+	 * @param principal the principal to use
+	 */
+	public void setPrincipal(Authentication principal) {
+		Assert.notNull(principal, "principal cannot be null");
+		this.authentication = () -> principal;
+	}
+
+	/**
+	 * Returns a {@link Consumer callback} that can be provided to
+	 * {@link org.springframework.web.client.RestClient.RequestHeadersSpec#httpRequest(Consumer)}
+	 * to make OAuth 2.0 requests by including the
+	 * {@link OAuth2AuthorizedClient#getAccessToken() access token} as a bearer token.
+	 *
+	 * <p>
+	 * This is useful for authorizing a client on a per-request basis, for example when
+	 * the {@code clientRegistrationId} is only known at runtime.
+	 *
+	 * <p>
+	 * Example usage:
+	 *
+	 * <pre>
+	 * RestClient restClient = RestClient.create();
+	 * ...
+	 * OAuth2ClientHttpRequestInterceptor requestInterceptor =
+	 *     new OAuth2ClientHttpRequestInterceptor(authorizedClientManager, clientRegistrationId);
+	 * String response = restClient.get()
+	 *     .uri(uri)
+	 *     .httpRequest(requestInterceptor.httpRequest())
+	 *     .retrieve()
+	 *     .onStatus(requestInterceptor.errorHandler())
+	 *     .body(String.class);
+	 * </pre>
+	 * @return a {@link Consumer} that can access the {@link ClientHttpRequest}
+	 * @see #errorHandler()
+	 */
+	public Consumer<ClientHttpRequest> httpRequest() {
+		return this::authorizeClient;
+	}
+
+	/**
+	 * Returns a {@link ResponseErrorHandler} that can be provided to
+	 * {@link org.springframework.web.client.RestClient.ResponseSpec#onStatus(ResponseErrorHandler)}
+	 * in order to forward authentication (HTTP 401 Unauthorized) and authorization (HTTP
+	 * 403 Forbidden) failures from an OAuth 2.0 Resource Server to a
+	 * {@link OAuth2AuthorizationFailureHandler}.
+	 *
+	 * <p>
+	 * This is useful for handling errors on a per-request basis, for example when the
+	 * {@code clientRegistrationId} is only known at runtime. See {@link #httpRequest()}
+	 * for more information.
+	 * @return the error handler
+	 * @see #httpRequest()
+	 */
+	public ResponseErrorHandler errorHandler() {
+		return new DefaultResponseErrorHandler() {
+			@Override
+			public void handleError(URI url, HttpMethod method, ClientHttpResponse response) throws IOException {
+				handleAuthorizationFailure(response.getHeaders(), response.getStatusCode());
+				super.handleError(url, method, response);
+			}
+		};
+	}
+
+	@Override
+	public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution)
+			throws IOException {
+		authorizeClient(request);
+		try {
+			ClientHttpResponse response = execution.execute(request, body);
+			handleAuthorizationFailure(response.getHeaders(), response.getStatusCode());
+			return response;
+		}
+		catch (RestClientResponseException ex) {
+			handleAuthorizationFailure(ex.getResponseHeaders(), ex.getStatusCode());
+			throw ex;
+		}
+		catch (OAuth2AuthorizationException ex) {
+			handleAuthorizationFailure(ex);
+			throw ex;
+		}
+	}
+
```

Comment:
Hey bro, I've tried to use ur interceptor in my production code, then I got an exception: IllegalArgumentException: URI with undefined scheme.
After some debugging attempts, I realized that the baseUrl, which I set here:
`return RestClient.builder ()
                .requestFactory (requestFactory)
                .baseUrl (commandCenterBaseUrl)
                .requestInterceptor (requestInterceptor)
                .build ();`
...didn't get passed to the interceptor... well Ok, I added an additional setter for the baseUrl and then...
```suggestion
@Override
    public ClientHttpResponse intercept (
            HttpRequest request, byte[] body, ClientHttpRequestExecution execution
    ) throws IOException {

        if (!isNullOrEmpty (baseUrl)){
            request = prependBaseUrl (request);
        }
        authorizeClient (request);

        try {
            ClientHttpResponse response = execution.execute (request, body);
            handleAuthorizationFailure (response.getHeaders (), response.getStatusCode ());
            return response;
        }
        catch (RestClientResponseException ex){
            handleAuthorizationFailure (ex.getResponseHeaders (), ex.getStatusCode ());
            throw ex;
        }
        catch (OAuth2AuthorizationException ex){
            handleAuthorizationFailure (ex);
            throw ex;
        }
    }

    private HttpRequest prependBaseUrl (HttpRequest request){
        URI uri = request.getURI ();
        URI completeUri = fromHttpUrl (baseUrl)
                .path (uri.getPath ())
                .query (uri.getQuery ())
                .build (true)
                .toUri ();

        return new HttpRequestWrapper (request){

            @Override
            public @NonNull URI getURI (){
                return completeUri;
            }
        };
    }
```
I hope this suggestion helps!

Reply:
Thanks for the suggestion @Lonelysoul-HayashiNoMirai! However, I'm not understanding your issue. This class should not be responsible for setting the `baseUrl`. Can you help me understand your issue better by providing a sample? **Note:** I have a TODO to add samples, so this will help immensely.

Reply:
> Thanks for the suggestion @Lonelysoul-HayashiNoMirai! However, I'm not understanding your issue. This class should not be responsible for setting the `baseUrl`. Can you help me understand your issue better by providing a sample? **Note:** I have a TODO to add samples, so this will help immensely.

I know that this class has nothing to do with "setting the baseUrl", then I should set it in here, right?
![image](https://github.com/user-attachments/assets/0cdc8a9d-042c-429a-81b0-10b5fd63b946)
Even after setting it in the rest client, I'm still getting this exception:
![image](https://github.com/user-attachments/assets/2d217416-05cc-4d81-847e-1ed684fc32d9)
When I was debugging, just before the exception was thrown, I clearly saw that the URL was missing the "baseUrl":
![Debug-Screenshot](https://github.com/user-attachments/assets/99d0c0d3-cdf8-4fb4-8b92-826e36d2841f)
With my modifications in the suggestion, everything is Ok again... that's all, bro

Reply:
@Lonelysoul-HayashiNoMirai thanks for reporting this issue and giving the interceptor a try. However, I am not able to reproduce your issue.

I have created a sample in a [branch](https://github.com/sjohnr/spring-security-samples/tree/gh-13588-rest-client) of the [Spring Security Samples](https://github.com/spring-projects/spring-security-samples) project that you can try ([code and instructions here](https://github.com/sjohnr/spring-security-samples/tree/gh-13588-rest-client/servlet/spring-boot/java/oauth2/httpclient)). Would you consider modifying that sample to reproduce the issue you're seeing and sharing it with me so I can investigate further?

Reply:
I should mention that in order to use the sample above, you will need to build this branch of spring-security and install it in your local maven cache. You can do so via the command `./gradlew publishToMavenLocal`.

Reply:
@sjohnr 
Thank bro for considering my feedback, really appreciate it...
However, I'm up to my neck in deadlines right now :(
Will check later if I have free time...

In the meantime, if u have some spare time, pls help me check this issue first (this may be the cause of the above issue): https://github.com/spring-projects/spring-security/pull/15437#issuecomment-2255630771